[GTK][WPE] Add a fallback mode for the DMABuf architecture

alexgcastro opened the pull request:
#### 7a3669210b66dee7b16f591ded098bb6424870b5
<pre>
[GTK][WPE] Add a fallback mode for the DMABuf architecture
<a href="https://bugs.webkit.org/show_bug.cgi?id=242108">https://bugs.webkit.org/show_bug.cgi?id=242108</a>

Reviewed by Žan Doberšek.

Adding a new fallback mode for the DMABuf architecture in Linux, for the
situations where DMABuf technology is not an option. This implementation
detects if DMABuf technology is supported and creates a GraphicsContextGLGBM
or a GraphicsContextGLFallback class accordingly. The selection of the
graphics context is implemented in createWebProcessGraphicsContextGL
function.

This change fixes the layout tests for the gtk port, because now it
detects the backend does not support DMABuf and uses the fallback
graphics context. This fallback context is slower because it copies the
memory to send it to the compositor thread but it is good enough for
testing situations.

* Source/WebCore/SourcesGTK.txt: Add the new GraphicsContextGLFallback
  and GraphicsContextGLANGLELinux classes.
* Source/WebCore/SourcesWPE.txt: Ditto.
* Source/WebCore/platform/TextureMapper.cmake: Added the
  GraphicsContextGLFallback header file.
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLANGLELinux.cpp:
  Added with the implementations of the GraphicsContextGLANGLE class
  shared by the linux graphics contexts.
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLFallback.cpp:
  Copied from Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapperANGLE.cpp.
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLFallback.h:
  Copied from Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapperANGLE.h.
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLGBM.cpp: Moved
  the implementation of the GraphicsContextGLANGLE functions to
  GraphicsContextGLANGLELinux.cpp.
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLGBM.h:
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLGBMTextureMapper.cpp:
* Source/WebCore/platform/graphics/nicosia/texmap/NicosiaGCGLANGLELayer.cpp:
(Nicosia::GCGLANGLELayer::swapBuffersIfNeeded): Reworked the
implementation of the fallback mode, we need to execute part of the code
in a lambda to make it run in the composition thread.
(Nicosia::GCGLANGLELayer::GCGLANGLELayer): Choose the implementation of
the graphics context we are going to  use.
* Source/WebCore/platform/graphics/nicosia/texmap/NicosiaGCGLANGLELayer.h:

Canonical link: <a href="https://commits.webkit.org/252208@main">https://commits.webkit.org/252208@main</a>
</pre>

zdobersek commented on Source/WebCore/platform/graphics/gbm/GraphicsContextGLGBM.cpp:
```diff
@@ -27,7 +27,7 @@
 #include "GraphicsContextGLGBM.h"
 
-#if ENABLE(WEBGL) && USE(LIBGBM) && USE(ANGLE)
+#if ENABLE(WEBGL) && USE(LIBGBM) && USE(ANGLE) && USE(TEXTURE_MAPPER_DMABUF)
```
USE(TEXTURE_MAPPER_DMABUF) is not meaningful here, this class doesn't use TextureMapper at all.

alexgcastro replied:
Thanks for the comments, we are reviewing the patch to add the runtime mode for the fallback backend directly in this patch.

zdobersek reviewed (commented):
USE_TEXTURE_MAPPER_DMABUF is an internal option about a TextureMapper-specific handler of DMABuf content. This won't be a public option at any point.

DMABuf support should also be detectable at runtime, and a suitable GraphicsContext3D class should be decided accordingly. This change doesn't handle that, so I don't know how much benefit there is to it.

`createWebProcessGraphicsContextGL()` should know what's possible, and then spawn a GraphicsContextGL instance based on that. So there should first be some separate GraphicsContextGL derivative that derives from GraphicsContextGLANGLE but does this fallback thing -- so a fallback GraphicsContextGL implementation. Once this exists, combined with the runtime DMABuf check, we can make the decision on-the-fly about what to use.

alexgcastro committed:
[GTK][WPE] Add a fallback mode for the DMABuf architecture
https://bugs.webkit.org/show_bug.cgi?id=242108

Reviewed by Žan Doberšek.

Adding a new fallback mode for the DMABuf architecture in Linux, for the
situations where DMABuf technology is not an option. This implementation
detects if DMABuf technology is supported and creates a GraphicsContextGLGBM
or a GraphicsContextGLFallback class accordingly. The selection of the
graphics context is implemented in createWebProcessGraphicsContextGL
function.

This change fixes the layout tests for the gtk port, because now it
detects the backend does not support DMABuf and uses the fallback
graphics context. This fallback context is slower because it copies the
memory to send it to the compositor thread but it is good enough for
testing situations.

* Source/WebCore/SourcesGTK.txt: Add the new GraphicsContextGLFallback
  and GraphicsContextGLANGLELinux classes.
* Source/WebCore/SourcesWPE.txt: Ditto.
* Source/WebCore/platform/TextureMapper.cmake: Added the
  GraphicsContextGLFallback header file.
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLANGLELinux.cpp:
  Added with the implementations of the GraphicsContextGLANGLE class
  shared by the linux graphics contexts.
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLFallback.cpp:
  Copied from Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapperANGLE.cpp.
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLFallback.h:
  Copied from Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapperANGLE.h.
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLGBM.cpp: Moved
  the implementation of the GraphicsContextGLANGLE functions to
  GraphicsContextGLANGLELinux.cpp.
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLGBM.h:
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLGBMTextureMapper.cpp:
* Source/WebCore/platform/graphics/nicosia/texmap/NicosiaGCGLANGLELayer.cpp:
(Nicosia::GCGLANGLELayer::swapBuffersIfNeeded): Reworked the
implementation of the fallback mode, we need to execute part of the code
in a lambda to make it run in the composition thread.
(Nicosia::GCGLANGLELayer::GCGLANGLELayer): Choose the implementation of
the graphics context we are going to  use.
* Source/WebCore/platform/graphics/nicosia/texmap/NicosiaGCGLANGLELayer.h:

Canonical link: https://commits.webkit.org/252208@main
